[xbmc][appmessenger] Add the ability to open dialogyesno by sending a threadmessage.

Paxxi opened the pull request:
So this is the continuation of the big appmessenger rewrite. Actually using it to get rid of our non-ui code calling directly into the ui and depending on it. Taking this one dialog at a time to keep the changes from ballooning and being unreviewable.

Paxxi committed:
Add the ability to open dialogyesno by sending a threadmessage.

Added two helper methods to make this easier in KODI::MESSAGING::HELPERS namespace

Changed all non-gui code to use this method instead to get rid of one gui include.

Paxxi commented:
@mkortstiege @MartijnKaijser and @Montellese  care to review?

xhaggi commented on xbmc/Application.cpp:
```diff
@@ -1551,8 +1550,11 @@ void CApplication::ReloadSkin(bool confirm/*=false*/)
+      options["heading"] = 13123;
+      options["text"] = 13111;
+      options["autoclose"] = 10000; //timeout value for autoclosing the dialog
+      if (CApplicationMessenger::Get().SendMsg(TMSG_GUI_DIALOG_YESNO, -1, -1, static_cast<void*>(&options)) != 1)
```
IMO this bloats the code massivly, why not introduce a helper class + method and do these stuff there?

Paxxi replied:
it does bloat the code, I'm not opposed to a helper but where to put it?

xhaggi replied:
should go into the same directory as the app mensenger something like `CApplicationMessageHelper::OpenDialogYesNo()` or similar.

xhaggi commented on xbmc/interfaces/Builtins.cpp:
```diff
@@ -1593,7 +1592,17 @@ int CBuiltins::Execute(const std::string& execString)
+      options["yeslabel"] = 20429;
+      options["nolabel"] = 20428;
+      
+      int result = CApplicationMessenger::Get().SendMsg(TMSG_GUI_DIALOG_YESNO, -1, -1, static_cast<void*>(&options));
```
does it make sense to introduce another overload of that method to get rid of the `-1, -1` parameters in the case we only need the payload parameter.

```
void SendMsg(uint32_t messageId, void* payload = nullptr);
```

Paxxi replied:
yeah I was thinking about it but it might cause trouble with overloading, need to look into it.

xhaggi replied:
sorry c/p and not checked, it's totally bullshit to default payload to nullptr. Without it works as it should.

Paxxi commented:
I tried some of the suggestions, @xhaggi  in the second commit I added helper methods for it.
I also modified SendMsg to avoid the static_cast everywhere but I don't think this one is a good idea even though it's very convenient, will lead to lots of methods in the binary.

xhaggi commented:
> I also modified SendMsg to avoid the static_cast everywhere but I don't think this one is a good idea even though it's very convenient, will lead to lots of methods in the binary.

Agree

Paxxi commented:
@mkortstiege I could use some xcode project updates
@xhaggi @Montellese I reworked everything to use the helpers instead and documented them.
Chose to stick them into a cpp file as there will likely be more helpers for the other dialogs once I get to them as well.

mkortstiege commented on Makefile.in:
```diff
@@ -60,6 +60,7 @@ DIRECTORY_ARCHIVES=$(DVDPLAYER_ARCHIVES) \
                    xbmc/main/posix/MessagePrinter.a \
                    xbmc/media/media.a \
                    xbmc/messaging/messaging.a \
+				   xbmc/messaging/helpers/messagingHelpers.a \
```
That looks wrong ;)

Paxxi replied:
Caught it and messed it up again, stupid makefile using tabs and spaces

MartijnKaijser replied:
still need to fix this ;)

Paxxi replied:
arghh, die make die

mkortstiege committed:
[osx/ios] xcode sync for PR7800

mkortstiege commented:
@Paxxi here your go. https://github.com/mkortstiege/xbmc/commit/944b64063720f74c49a20d722d0e8a03aa744f4a

xhaggi commented on xbmc/messaging/helpers/DialogHelper.h:
```diff
@@ -0,0 +1,80 @@
+  \sa ShowYesNoDialogLines
+  \sa CGUIDialogYesNo::ShowAndGetInput
+*/
+int ShowYesNoDialogText(CVariant heading, CVariant text, CVariant noLabel = "", CVariant yesLabel = "", uint32_t autoCloseTimeout = 0);
```
why do we use functions instead of a helper class with static methods? What is the benefit?

Paxxi replied:
There's no added benefit of wrapping them in a class, they're already encapsulated by the namespace.
A class with static methods is just used for grouping and naming and a namespace does the same in this case as we don't keep any state

xhaggi replied:
The biggest advantage over functions + namespace is that you can use overloads and IMO a cleaner design. @Montellese what do you think?

Paxxi replied:
There's nothing preventing using overloads here but I chose this naming specifically to avoid the overload hell we currently have where it's really easy to call the wrong showandgetinput

xhaggi replied:
this is because of the defaults we use in mix with the overloads, but okay leave it.

FernetMenta commented on xbmc/dialogs/GUIDialogYesNo.h:
```diff
@@ -90,6 +90,27 @@ class CGUIDialogYesNo :
+                        To avoid spelling errors it's advised to use
+                        the static strings in KODI::MESSAGING::HELPERS
+
+    \returns -1 for cancelled, 0 for No and 1 for Yes
```
if those were defines, the code would be more readable

Paxxi replied:
Ahh they are but I didn't update the help text, will fix it

tamland replied:
Or better: bool, or if cancelled is really needed: enum. Returning int here makes no sense

Paxxi replied:
It's an int because appmessenger is general purpose, the helper methods could off define and use an enum

FernetMenta commented on xbmc/pvr/PVRManager.cpp:
```diff
@@ -1506,13 +1506,11 @@ bool CPVRManager::CanSystemPowerdown(bool bAskUser /*= true*/) const
-                                                   CVariant{222},   // "Cancel"
-                                                   CVariant{19696}, // "Shutdown anyway"
-                                                   10000); //This is a timeout and IS NOT a CVariant
+        bReturn = 1 == HELPERS::ShowYesNoDialogText(CVariant{19685}, // "Confirm shutdown"
```
brackets highly welcome.

Paxxi replied:
Will add

tamland commented on xbmc/messaging/helpers/DialogHelper.cpp:
```diff
@@ -0,0 +1,58 @@
+{
+namespace HELPERS
+{
+int ShowYesNoDialogText(CVariant heading, CVariant text, CVariant noLabel, CVariant yesLabel, uint32_t autoCloseTimeout)
```
This is.. weirdly named. Should name it so it convey what it's *actually* doing. The original `CGUIDialogYesNo::ShowAndGetInput` much clearer in that regard.

tamland commented:
Excuse my ignorance if I missed something, but is there a reason the message payload here have to be a Variant + all those key defines instead of a regular POD?

Paxxi commented:
@tamland it was first designed to be without helper methods so I aimed to minimize extra structs for the message. With this design it might make more sense to go with a struct

Paxxi commented:
jenkins build this please

Paxxi commented:
jenkins build this please

Paxxi commented:
jenkins build this please

Ok I'm thinking this is in ok shape now so unless there's some big objections I'll merge this in a day or two and continue with the rest of the dialogs.

Paxxi commented:
jenkins build this please
